### PR TITLE
[BACKLOG-17675] audit entry for remote execution

### PR DESCRIPTION
**Pull request by chiltonj**
@pentaho/rebelalliance @pentaho/rogueone 

Adds audit entry for start / finish. Since the same id is used for the complete audit trail the duration per item is cumulative -- which is the accurate duration.

### CODE REVIEW

**Comment by pentaho-amessier** [comment]
Wouldn't it make more sense to put this code directly inside actionInvoker.invokeAction? Otherwise the effort will need to be duplicated in every "entry point" to the action invoking code, which at the moment is the ActionResource (for long running / marathon) and ActionIntokerSystemListener (for short running / chronos)

**Comment by chiltonj** [comment]
@pentaho-amessier Does the actionInvoker.invokeAction have multiple implementations? If so are they   called out side of a WN context ? However, as you point out It may be necessary to add similar audit to chronos. But I am not going to push it down into common code that is executed by the server as well.

**Comment by pentaho-amessier** [comment]
@chiltonj there are multiple implementations, but it all points down to a common implementation in one of the classes - currently all various implementations end up calling the same code, if I recall, although clearly we cannot rely on this always being the case. Given that this should only apply to WN, you are right not to push it to the common code. In that case, perhaps we can put the makeAuditRecord and any other pieces that make sense in a common location and add the logic to ActionInvokerSystemListener... Or if you think it makes sense, pernahs we can create a common entry point for both ActionResource and ActionInvokerSystemListener, so that the call this common point instead of the actionInvoker directly and then we could put the auditing code in one place. Whatever you think is best.

**Comment by chiltonj** [comment]
@pentaho-amessier your point is valid -- I will add in a decorator like thing that can be used in WN context

**Comment by chiltonj** [comment]
Changes made

File: extensions/src/main/java/org/pentaho/platform/web/http/api/resources/WorkerNodeActionInvokerAuditor.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.pentaho.platform.engine.core.audit.AuditHelper;
+import org.pentaho.platform.engine.core.audit.MessageTypes;
+import org.pentaho.platform.engine.core.system.PentahoSessionHolder;
+import org.pentaho.platform.util.ActionUtil;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+public class WorkerNodeActionInvokerAuditor implements IActionInvoker {
```
Comment (pentaho-amessier) [inline extensions/src/main/java/org/pentaho/platform/web/http/api/resources/WorkerNodeActionInvokerAuditor.java]: A short description explaining the purpose of this class would be helpful

File: extensions/src/test/java/org/pentaho/platform/web/http/api/resources/ActionResourceTest.java [hunk 2]
```diff
@@ -192,8 +192,30 @@ public void testRunnableRun() throws Exception {
...
+
+    final WorkerNodeActionInvokerAuditor wnActionInvokerAuditer = new WorkerNodeActionInvokerAuditor( actionInvoker );
+    WorkerNodeActionInvokerAuditor wnaiu = Mockito.spy( wnActionInvokerAuditer );
+    wnaiu.invokeAction( actionMock, actionUser, actionMapMock );
+
+    Mockito.verify( actionInvoker, Mockito.times( 1 ) ).invokeAction(  Mockito.eq( actionMock ), Mockito.eq( actionUser ),  Mockito.eq( actionMapMock ) );
+
+    Mockito.verify( wnaiu, Mockito.times( 2 ) ).makeAuditRecord( Mockito.anyLong() ,Mockito.anyString(), Mockito.anyMap() ,  Mockito.anyString() );
+
+  }}
```
Comment (pentaho-amessier) [inline extensions/src/test/java/org/pentaho/platform/web/http/api/resources/ActionResourceTest.java]: Is this closing bracket in the right place?

Comment (chiltonj) [inline extensions/src/test/java/org/pentaho/platform/web/http/api/resources/ActionResourceTest.java]: Done: fixed a bunch of checkstyle atrocities

File: extensions/src/main/java/org/pentaho/platform/web/http/api/resources/WorkerNodeActionInvokerAuditor.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    try {
+      return actionInvoker.invokeAction( action, user, params );
+    } finally {
+      makeAuditRecord( stopWatch.getTime() / 1000, MessageTypes.INSTANCE_END, auditParams, action.getClass().getName() );
+    }
+  }
+
+
+  protected void makeAuditRecord( final float time, final String messageType,
+                                 final Map<String, Serializable> actionParams, String className ) {
```
Comment (pentaho-mzur) [inline extensions/src/main/java/org/pentaho/platform/web/http/api/resources/WorkerNodeActionInvokerAuditor.java]: add 'final' to className?

**Pull request merged**